### PR TITLE
Correctly handle quoted console options and whitespace

### LINKED ISSUE #15563: Tab completing breaks quotes/spaces of options
## Steps to reproduce

Perform tab completion where the `<caret>` is

```
use admin/postgres/postgres_sql
run password='a b c'      usern<caret>
```

### Current output
```
run password=a b c username=
```

### Expected output

```
run password='a b c'      username=
```

### PR DESCRIPTION
**Pull request by adfoster-r7**
Correctly handles quoted console options and whitespace. Closes https://github.com/rapid7/metasploit-framework/issues/15563

### Before

When the user performs tab completion, the entire string is entirely reconstructed, which can generate invalid commands, as this will get incorrectly parsed as `["run", "password=a", "b", "c", "username="]`:

Input:
```
use admin/postgres/postgres_sql
run password='a b c'      usern<tab>
```

Result:
```
run password=a b c username=<caret>
```

### After

Only the tab completed word is replaced.

Input:
```
use admin/postgres/postgres_sql
run password='a b c'      usern<tab>
```

Result:
```
run password='a b c'      username=<caret>
```

One new edgecase spotted, is tab completing option values with spaces already present will now continually autocomplete whitespace

```
set smbpass          <tab>
```

## Verification

Replicate the above example on master versus this branch., and tab complete additional scenarios that might represent edge cases.

Test: The alias plugin works as expected, and tab completion for windows/linux via meterpreter, specifically with paths with spaces, ensure that these verification steps still work https://github.com/rapid7/metasploit-framework/pull/15050

### CODE REVIEW

File: lib/rex/ui/text/dispatcher_shell.rb [hunk 1]
```diff
@@ -662,51 +669,32 @@ def unblock_command(cmd)
   # This code was originally taken from https://github.com/ruby/ruby/blob/93420d34aaf8c30f11a66dd08eb186da922c831d/lib/shellwords.rb#L88
   #
   def shellsplitex(line)
-    quote = nil
-    words = []
+    unbalanced_quote = nil
+    tokens = []
     field = String.new
-    regexp = %r{
-      \G\s*(?>(?<word>[^\s\'\"]+)             # Words within str
```
Comment (adfoster-r7) [inline lib/rex/ui/text/dispatcher_shell.rb:669]: These comments were added by https://github.com/rapid7/metasploit-framework/pull/15050, but the changes broke the regex. I've reverted it to the original regex, but now I'm additionally matching preceding whitespace to correctly track the token index

File: lib/rex/ui/text/dispatcher_shell.rb [hunk 2]
```diff
@@ -662,51 +669,32 @@ def unblock_command(cmd)
...
-        if quote.nil?
-          quote = garbage
-        else
-          field << garbage
-        end
-        next
+        quote_start_begin = Regexp.last_match.begin(0)
+        unbalanced_quote = garbage
+        field << line[quote_start_begin + 1..-1]
+        break
```
Comment (adfoster-r7) [inline lib/rex/ui/text/dispatcher_shell.rb:685]: Now if there's garbage: Instead of continuing to parse the remaining string, we just return a single token with the remaining value

**Comment by dwelch-r7** [comment]
Got a couple of examples of things not quite working right:

In Meterpreter on windows
`dir \<tab>` shows options for tab completion but all fail with "[-] stdapi_fs_stat: Operation failed: The system cannot find the file specified." since I believe you need to ebgin with `C:\` Not sure if that falls within the scope here though

`dir C:\<tab>` displays a list of tab complete options but typing the next letter onf one of them and hitting tab results in nothing

`dir C:\\<tab>` replaces the `\\` with `\`

`dir C:\\Pro<tab>` will tab complete to `C:\Program` and now <tab> no longer does anything since the `\\` was again replaced with `\`

`dir 'Program Files '<tab>` autocompletes to `dir Program\ Files\ (x86)` and I realise no one would do this on purpose but I just hit <tab> out of habit and well....y'know

This one is more of a UX thing but anyway `cd fold<tab>` goes to `cd folder ` note the space after the folder name which makes continuing a bit of an annoyance if you want to traverse multiple folders at a time for example

File: spec/lib/msf/ui/text/dispatcher_shell_spec.rb [hunk 3]
```diff
@@ -1,32 +1,264 @@
...
   let(:subject) do
     dummy_class = Class.new
     dummy_class.include described_class
     dummy_class.new(prompt, prompt_char)
   end
 
+  let(:mock_dispatcher) do
+    dispatcher = double :mock_dispatcher
+    allow(dispatcher).to receive(:tab_complete_helper) do |_current_word, _preceding_words|
+      ['username=']
```
Comment (dwelch-r7) [inline spec/lib/msf/ui/text/dispatcher_shell_spec.rb:16]: shouldn't `password=` be in here aswell since it's an option that _should_ be available for completion in the tests below?

File: spec/lib/msf/ui/text/dispatcher_shell_spec.rb [hunk 4]
```diff
@@ -1,32 +1,264 @@
...
+      ]
+    end
+
+    before(:each) do
+      allow(subject).to receive(:dispatcher_stack).and_return(dispatcher_stack)
+    end
+
+    [
+      { input: '', expected: nil },
+      { input: '      ', expected: ['      username='] },
```
Comment (dwelch-r7) [inline spec/lib/msf/ui/text/dispatcher_shell_spec.rb:34]: reading the tests and seeing this _not_ try to suggest `password=` confused me a little since it's used in the rest of the tests, also be nice to test completion if more than one option is available

**Comment by dwelch-r7** [comment]
Tried to put together some inputs/expected outputs for my previous comment, I'm not sure how I feel about all of them especially the ones with a beginning `'` but not tab completing the final one so we can still traverse folders :shrug: 
```
{ input: 'cd folder', expected: ['cd folder\'] },
{ input: "cd 'folder", expected: ["cd 'folder\"] },
{ input: "cd 'folder", expected: ["cd 'folder\"] },
{ input: "cd 'folder\", expected: ["cd 'folder\folder2"] },
{ input: "cd 'folder", expected: ["cd 'folder\"] },
{ input: "cd C:\\", expected: ["cd C:\\"] }, # plus all the suggestions? not sure how to show that here
{ input: "cd C:\\D", expected: ["cd C:\\Documents\ and\ Settings\"] },
{ input: "cd 'C:\\D", expected: ["cd 'C:\\Documents and Settings\"] }, 
```

**Comment by adfoster-r7** [comment]
Also noticed that it fixes tab completion for targets on the postgres_payload module.

On master, this tab completion doesn't do anything:
```
use linux/postgres/postgres_payload
set target Linux\ x86
```

On this branch, tab completion works as expected:
```
msf6 exploit(linux/postgres/postgres_payload) > set target Linux\ x86
set target Linux\ x86     set target Linux\ x86_64
```

**Comment by dwelch-r7** [comment]
# Release Notes

This PR implements a fix to correctly handle quoted console options and whitespace
